frame/im-online/src/lib.rs

```diff
@@ -365,156 +423,128 @@ impl<T: Trait> Module<T> {
+			return Err(OffchainErr::AlreadyOnline(authority_index));
 		}
 
-		// fail only after trying all keys.
-		results.into_iter().collect::<Result<Vec<_>, OffchainErr>>()?;
-
-		// once finished we set the worker status without comparing
-		// if the existing value changed in the meantime. this is
-		// because at this point the heartbeat was definitely submitted.
-		Self::set_worker_status(block_number, true);
+		// acquire lock for that authority at current heartbeat to make sure we don't
+		// send concurrent heartbeats.
+		Self::with_heartbeat_lock(
+			authority_index,
+			session_index,
+			block_number,
+			|| {
+				let call = prepare_heartbeat()?;
+				debug::info!(
+					target: "imonline",
+					"[index: {:?}] Reporting im-online at block: {:?} (session: {:?}): {:?}",
+					authority_index,
+					block_number,
+					session_index,
+					call,
+				);
 
-		Ok(())
-	}
+				T::SubmitTransaction::submit_unsigned(call)
+					.map_err(|_| OffchainErr::SubmitTransaction)?;
 
-	fn compare_and_set_worker_status(
-		gossipping_at: T::BlockNumber,
-		done: bool,
-		curr_worker_status: Option<Vec<u8>>,
-	) -> bool {
-		let enc = WorkerStatus {
-			done,
-			gossipping_at,
-		};
-		sp_io::offchain::local_storage_compare_and_set(
-			StorageKind::PERSISTENT,
-			DB_KEY,
-			curr_worker_status,
-			&enc.encode()
+				Ok(())
+			},
 		)
 	}
 
-	fn set_worker_status(
-		gossipping_at: T::BlockNumber,
-		done: bool,
-	) {
-		let enc = WorkerStatus {
-			done,
-			gossipping_at,
-		};
-		sp_io::offchain::local_storage_set(StorageKind::PERSISTENT, DB_KEY, &enc.encode());
+	fn local_authority_keys() -> impl Iterator<Item=(u32, T::AuthorityId)> {
+		// we run only when a local authority key is configured
+		let authorities = Keys::<T>::get();
+		let mut local_keys = T::AuthorityId::all();
+		local_keys.sort();
+
+		authorities.into_iter()
+			.enumerate()
+			.filter_map(move |(index, authority)| {
+				local_keys.binary_search(&authority)
+					.ok()
+					.map(|location| (index as u32, local_keys[location].clone()))
+			})
 	}
 
-	// Checks if a heartbeat gossip already occurred at this block number.
-	// Returns a tuple of `(current worker status, bool)`, whereby the bool
-	// is true if not yet gossipped.
-	fn check_not_yet_gossipped(
+	fn with_heartbeat_lock<R>(
+		authority_index: u32,
+		session_index: SessionIndex,
 		now: T::BlockNumber,
-		next_gossip: T::BlockNumber,
-	) -> Result<(Option<Vec<u8>>, bool), OffchainErr> {
-		let last_gossip = sp_io::offchain::local_storage_get(StorageKind::PERSISTENT, DB_KEY);
-		match last_gossip {
-			Some(last) => {
-				let worker_status: WorkerStatus<T::BlockNumber> = Decode::decode(&mut &last[..])
-					.map_err(|_| OffchainErr::DecodeWorkerStatus)?;
-
-				let was_aborted = !worker_status.done && worker_status.gossipping_at < now;
+		f: impl FnOnce() -> OffchainResult<T, R>,
+	) -> OffchainResult<T, R> {
+		let key = {
+			let mut key = DB_PREFIX.to_vec();
+			key.extend(authority_index.encode());
```

Comment:
A heads up here that this will pollute local storage a bit, since the keys are based on `authority_index` and we don't really remove the locks at any point in the future.
However this should always be bounded to the maximal size of validator set (which I believe has some reasonable values anyway). So I don't think it's necessary to iterate and clear it, if we decide it is, it can be easily added retroactively.